Network/HTTP2/Server/Worker.hs

```diff
@@ -161,11 +163,28 @@ sendStreaming Context{..} strm th strmbdy = do
                         T.resume th
                     , outBodyPushFinal = \b -> do
                         T.pause th
-                        atomically $ writeTBQueue tbq (StreamingBuilder b (EndOfStream Nothing))
+                        atomicWriteIORef finishedOrCancelled True
+                        atomically $ writeTBQueue tbq $ StreamingBuilder b (EndOfStream Nothing)
+                        atomically $ writeTBQueue tbq $ StreamingFinished Nothing
                         T.resume th
                     , outBodyFlush = atomically $ writeTBQueue tbq StreamingFlush
+                    , outBodyCancel = \mErr -> do
+                        -- See comment in Network.HTTP2.Client.Run
+                        headersResult <- readMVar (streamHeadersSent strm)
+                        case headersResult of
+                          Left _ ->
+                            return ()
+                          Right () -> do
+                            already <- atomicModifyIORef finishedOrCancelled (\x -> (True, x))
+                            if already then
+                              return ()
+                            else
+                              atomically $ writeTBQueue tbq (StreamingCancelled mErr)
                     }
-            finished = atomically $ writeTBQueue tbq $ StreamingFinished Nothing
+            finished = do
+              already <- atomicModifyIORef finishedOrCancelled (\x -> (True, x))
+              unless already $ do
+                atomically $ writeTBQueue tbq $ StreamingFinished Nothing
```

Comment:
Same comments here as in the client side (let's make these single STM transactions). It would be nice if we could somehow reduce this duplication :thinking: